lindi/LindiH5pyFile/LindiH5pyFile.py

```diff
@@ -0,0 +1,218 @@
+from .LindiH5pyDataset import LindiH5pyDataset
+from .LindiH5pyAttributes import LindiH5pyAttributes
+from .LindiH5pyReference import LindiH5pyReference
+
+
+class LindiH5pyFile(h5py.File):
+    def __init__(self, _file_object: Union[h5py.File, zarr.Group]):
+        """
+        Do not use this constructor directly. Instead, use:
+        from_reference_file_system, from_zarr_store, from_zarr_group,
+        or from_h5py_file
+        """
+        self._file_object = _file_object
+        self._the_group = LindiH5pyGroup(_file_object, self)
+
+    @staticmethod
+    def from_reference_file_system(rfs: Union[dict, str]):
+        """
+        Create a LindiH5pyFile from a reference file system.
+        """
+        if isinstance(rfs, str):
+            if rfs.startswith("http") or rfs.startswith("https"):
+                with tempfile.TemporaryDirectory() as tmpdir:
+                    filename = f"{tmpdir}/temp.zarr.json"
+                    _download_file(rfs, filename)
+                    with open(filename, "r") as f:
+                        data = json.load(f)
+                    assert isinstance(data, dict)  # prevent infinite recursion
+                    return LindiH5pyFile.from_reference_file_system(data)
+            else:
+                with open(rfs, "r") as f:
+                    data = json.load(f)
+                assert isinstance(data, dict)  # prevent infinite recursion
+                return LindiH5pyFile.from_reference_file_system(data)
+        elif isinstance(rfs, dict):
+            fs = ReferenceFileSystem(rfs).get_mapper(root="")
+            return LindiH5pyFile.from_zarr_store(fs)
+        else:
+            raise Exception(f"Unhandled type for rfs: {type(rfs)}")
+
+    @staticmethod
+    def from_zarr_store(zarr_store: Union[ZarrStore, FSMap]):
+        """
+        Create a LindiH5pyFile from a zarr store.
+        """
+        zarr_group = zarr.open(store=zarr_store, mode="r")
+        assert isinstance(zarr_group, zarr.Group)
+        return LindiH5pyFile.from_zarr_group(zarr_group)
+
+    @staticmethod
+    def from_zarr_group(zarr_group: zarr.Group):
+        """
+        Create a LindiH5pyFile from a zarr group.
+        """
+        return LindiH5pyFile(zarr_group)
+
+    @staticmethod
+    def from_h5py_file(h5py_file: h5py.File):
+        """
+        Create a LindiH5pyFile from an h5py file.
+        """
+        return LindiH5pyFile(h5py_file)
+
+    @property
+    def attrs(self):  # type: ignore
+        if isinstance(self._file_object, h5py.File):
+            attrs_type = 'h5py'
+        elif isinstance(self._file_object, zarr.Group):
+            attrs_type = 'zarr'
+        else:
+            raise Exception(f'Unexpected file object type: {type(self._file_object)}')
+        return LindiH5pyAttributes(self._file_object.attrs, attrs_type=attrs_type)
+
+    @property
+    def filename(self):
+        # This is not a string, but this is what h5py seems to do
```

Comment:
```suggestion
        # This is not a string, but this is what h5py seems to do
```
I don't understand this comment. `self._file_object.filename` returns a string as far as I can tell.

Reply:
In the case where h5py.File was opened with a file-like object, I think this returns an object, not a string.

Reply:
Gotcha